### ISSUE TITLE
Kestrel server does not get grade A in SSL tests because it has no cipher suit preference

### ISSUE THREAD

**Issue by Tratcher** [dotnet/corefx#40830]
_From @ChrML on Thursday, September 5, 2019 7:09:29 AM_

### Description
From ASP Core 3.0 preview6 and newer it is now possible to get a grade B, instead of a previous grade F, on the SSL test (https://www.ssllabs.com/ssltest/). Because from 3.0 the insecure cipher suits are now disabled by default.

But still in 3.0-preview8, the Kestrel server does not have any cipher suit preference. This caps the SSL test grade to B instead of A. Grade A servers are regarded the most secure servers.

### Describe the solution you'd like
The server should prefer the "Perfect Forward Secrecy" compliant cipher suits (ECDHE) over the weaker cipher suits.

### Additional context
Here is the SSL report from the cipher suit chapter (on a 3.0-preview8 application):

> # Cipher Suites
> ## TLS 1.2 (server has no preference)
> TLS_RSA_WITH_AES_128_CBC_SHA (0x2f) WEAK 128
> TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (0xc013) ECDH secp521r1 (eq. 15360 bits RSA) FS WEAK 128
> TLS_RSA_WITH_AES_128_CBC_SHA256 (0x3c) WEAK 128
> TLS_RSA_WITH_AES_128_GCM_SHA256 (0x9c) WEAK 128
> TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256 (0xc027) ECDH secp521r1 (eq. 15360 bits RSA) FS WEAK 128
> TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (0xc02f) ECDH secp521r1 (eq. 15360 bits RSA) FS 128
> TLS_RSA_WITH_AES_256_CBC_SHA (0x35) WEAK 256
> TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (0xc014) ECDH secp521r1 (eq. 15360 bits RSA) FS WEAK 256
> TLS_RSA_WITH_AES_256_CBC_SHA256 (0x3d) WEAK 256
> TLS_RSA_WITH_AES_256_GCM_SHA384 (0x9d) WEAK 256
> TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384 (0xc028) ECDH secp521r1 (eq. 15360 bits RSA) FS WEAK 256
> TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (0xc030) ECDH secp521r1 (eq. 15360 bits RSA) FS 256
> TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 (0xcca8) ECDH secp521r1 (eq. 15360 bits RSA) FS 256

Here is the description from SSL labs why the Kestrel server is capped to grade B:

> ## Penalty for not using forward secrecy (B)
> Forward secrecy (FS) also known as perfect forward secrecy (PFS), is a property of secure communication protocols in which compromises of long-term keys does not compromise past session keys. Forward secrecy protects past sessions against future compromises of private key. The very popular RSA key exchange doesn’t provide forward secrecy. You need to support and prefer ECDHE suites in order to enable forward secrecy with modern web browsers.

> SSL Labs will start penalizing servers that don’t support forward secrecy; Grade will be capped to B. We will not penalize sites that use suites without forward secrecy provided they are never negotiated with clients that can do better.

Simply making the server prefer the existing ECDHE cipher suits over the weak cipher suits will suffice for getting a grade A.

_Copied from original issue: aspnet/AspNetCore#13700_

**Comment by Tratcher** [comment]
_From @Tratcher on Thursday, September 5, 2019 10:16:26 AM_

Was this Windows, Linux, or Mac? You'll get different results because Kestrel does not control this ordering, it's controlled by the operating system (or OpenSSL for Linux).

**Comment by Tratcher** [comment]
_From @ChrML on Thursday, September 5, 2019 10:36:44 AM_

> Was this Windows, Linux, or Mac? You'll get different results because Kestrel does not control this ordering, it's controlled by the operating system (or OpenSSL for Linux).

Hosted by Linux under this official Docker- image: https://hub.docker.com/_/microsoft-dotnet-core-aspnet (mcr.microsoft.com/dotnet/core/aspnet:3.0)

We have not really touched the OS at all, just copied the application to inside the official Docker image and deployed it to a Kubernetes cluster.

**Comment by Tratcher** [comment]
_From @Tratcher on Thursday, September 5, 2019 12:15:13 PM_

I'll forward this to the SslStream owners to see if it's configurable in their OpenSsl usage.

**Comment by ChrML** [comment]
Thanks. Following.

**Comment by stephentoub** [comment]
cc: @krwq

**Comment by davidsh** [comment]
@bartonjs

**Comment by krwq** [comment]
AFAIR OpenSSL and SecureTransport should respect sorting when using [CipherSuitePolicy](https://github.com/dotnet/corefx/blob/master/src/System.Net.Security/ref/System.Net.Security.cs#L141). The TLS implementations might re-sort the list and put TLS1.3 cipher suites first though (which should still grade it as A).

If you prefer to use default settings with "prefer forward secrecy" I'd need to investigate if there is a direct option for that anywhere (can't find by quick looking) and if not we can likely emulate it using CipherSuitePolicy (i.e. read the OS defaults and provide it as a list) but we'd need API proposal for that.

One thing to try is to upgrade your OpenSSL to latest and see if the settings improved.

**Comment by ChrML** [comment]
The server has no preference at all regarding cipher suit ordering, it only lists which cipher suits it supports (with PFS suits being in the list, but not preferred by the server).

Do you mean that adding a CipherSuitePolicy to the code with the wanted cipher suits as a manually sorted list, should make the server prefer them in that order? I was under the impression that CipherSuitePolicy only changed the supported suites, but didn't enable server preference. I need to test this. If that is the case, it should probably be added to the documentation. 

I think that the server by default should prefer the most secure cipher suits though. We use the ASP Core 3.0 docker image, not sure which version of OpenSSL is included in that image.

**Comment by krwq** [comment]
@ChrML, yes specifying CipherSuitePolicy manually will do best attempt to respect the given order (both on server and client) but order it is not guaranteed since it's underlying implementation which eventually decides.

SSL_OP_CIPHER_SERVER_PREFERENCE setting will not be changed though with that so I believe it's client order which wins by default (at least with OpenSSL, unsure with SecureTransport) - the only exception to that is TLS 1.3 - if both sides negotiate any TLS 1.3 cipher suites the TLS 1.3 cipher suite will be preferred.

> I think that the server by default should prefer the most secure cipher suits though.

Well, this is rather arbitrary - some will say ChaCha20 is better than AES some claim otherwise.

Also note that cipher suites are sent plaintext in the client hello which implies that MITM is able to degrade chosen cipher suites to the weakest possible and if degradation leads to breaking bad crypto then Finished message which contains MAC of the previous messages (including client hello) can also be forged so choosing "best" one will not give you much. You should trust all of cipher suites you set on the server and client and if you're not comfortable with any of them being chosen you should remove it from the list.

**Comment by ChrML** [comment]
@krwq Yes, client order will win by default. This does not matter though, because the SSL Cert rating for the server will only consider what the server prefers. There is a seperate SSL Cert test for clients which considers client ordering, but that is outside the scope of what a server application developer can control.

What is important is that the server has a cipher suit preference. And that the servers cipher suits preferred by the server support PFS and is not a "broken" cipher suit (eg proven insecure). This is all that is required to get a A grade rating for the server. Whether this cipher suit will actually be negotiated does not matter, and is also up to the client.

Downgrade attacks is the reason why the server is not allowed to support insecure cipher suits. If the server supports insecure cipher suits (cracked), the SSL tests will automatically degrade the server to grade F. Because the client can be "downgrade-attacked" to this insecure suite by a man in the middle. This can be perfectly handled by the server removing these cipher suits from CipherSuitePolicy as soon as they get cracked.

Default cipher suits in the ASP Core 3.0 docker image seems up to date and has good enough cipher suits to get a grade B. All that is required to get grade A is that the server prefers a cipher suit that is considered "perfect forward secret". At the moment the server has no preference and leaves it up to the client.

(yes, you can get "downgrade-attacked" to a weaker cipher suit that is not PFS, but that is accepted by SSL test and will not cap the server's grade).

**Comment by davidsh** [comment]
@krwq @Tratcher Is there something actionable here for .NET Core to make this scenario work? Or does this scenario require OS configuration (i.e. env variables) etc.? Trying to decide what to do with this issue.

**Comment by krwq** [comment]
I think the only actionable thing at this moment might be adding something equivalent to CipherSuitesPolicy.PreferServerOrder but there would have to be some investigation done prior to that as there are some rough edges (i.e. how to express `use OS default cipher suites list` since this is currently done by passing `(CipherSuitesPolicy)null` and check if this is possible to implement on at least OSX)

**Comment by ChrML** [comment]
In case other people are having the same issue and need a workaround, here is one temporary solution:

Adding this piece of code to `Program.cs` configures the Kestrel server for TLS 1.2 and TLS 1.3 only with secure cipher suits. CipherSuitesPolicy only works on non-Windows operating systems.


		.ConfigureKestrel((context, options) =>
		{
			options.ConfigureHttpsDefaults(httpsOptions =>
			{
				// Important: Only enable TLS 1.2 and TLS 1.3 to comply with SSL Server tests.
				//            TLS 1.1, TLS 1.0 and SSLv3 are considered insecure by todays standards.
				httpsOptions.SslProtocols = System.Security.Authentication.SslProtocols.Tls12 | System.Security.Authentication.SslProtocols.Tls13;

		// Configure HTTPS certificate.
		if (!String.IsNullOrEmpty(httpsCertificateFile))
		{
			httpsOptions.ServerCertificate = new X509Certificate2(httpsCertificateFile);
		}

		// Configure the cipher suits preferred and supported by the server. (Windows- servers are not so keen on doing this ...)
		if (!System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform.Windows))
		{
			httpsOptions.OnAuthenticate = (conContext, sslAuthOptions) =>
			{
				sslAuthOptions.CipherSuitesPolicy = new System.Net.Security.CipherSuitesPolicy(
					new System.Net.Security.TlsCipherSuite[]
					{
						// Cipher suits as recommended by: https://wiki.mozilla.org/Security/Server_Side_TLS
						// Listed in preferred order.

						// Highly secure TLS 1.3 cipher suits:
						System.Net.Security.TlsCipherSuite.TLS_AES_128_GCM_SHA256,
						System.Net.Security.TlsCipherSuite.TLS_AES_256_GCM_SHA384,
						System.Net.Security.TlsCipherSuite.TLS_CHACHA20_POLY1305_SHA256,

						// Medium secure compatibility TLS 1.2 cipher suits:
						System.Net.Security.TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
						System.Net.Security.TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
						System.Net.Security.TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
						System.Net.Security.TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
						System.Net.Security.TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
						System.Net.Security.TlsCipherSuite.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
						System.Net.Security.TlsCipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
						System.Net.Security.TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
					}
					);
					};
				}
			});
		});

This code yields the following SSL Labs results:
![Summary1](https://user-images.githubusercontent.com/38761430/65313996-b08ab080-db95-11e9-9ed2-e727727f858c.PNG)

![Summary2](https://user-images.githubusercontent.com/38761430/65314004-b4b6ce00-db95-11e9-9549-489a9ef443db.PNG)

**NOTE:** For some reason the legacy cipher-suits for TLS 1.2 does not get applied on Linux under the Docker-image `mcr.microsoft.com/dotnet/core/aspnet:3.0`.

This may cause older clients to be unable to connect. Such as Windows 7 / 8.1 machines with IE11 and Safari 8 and older browsers will stop working. IE/Safari on newer OS and Chrome/Firefox regardless of OS is still fine.

**Comment by krwq** [comment]
@ChrML to avoid allocations might be worth storing that CipherSuitesPolicy instance somewhere - otherwise it will be constructed on every request

**Comment by ChrML** [comment]
@krwq Good point, thanks! Did not realize it's called for every request.

**Comment by karelz** [comment]
@krwq what do you think needs to happen here as next step? Can you please help us clarify?

**Comment by krwq** [comment]
@karelz this request is to add option to pick between server or client cipher suite preference. So possibly something like:

```class
public partial class SslServerAuthenticationOptions
{
  publlic bool PreferServerOrder { get; set; } // i.e. on OpenSSL that would set SSL_OP_CIPHER_SERVER_PREFERENCE
}
```

Having said that I do not believe this feature will add any security to any system (it will always be as good as your weakest cipher suite) but it does currently give you better rank on some websites (like mentioned above https://www.ssllabs.com/ssltest/)

**Comment by karelz** [comment]
@Tratcher how much do you want this for Kestrel?
cc @samsp-msft

**Comment by Tratcher** [comment]
@bartonjs, @anurse can you share your notes from the crypto-board meeting about this?

**Comment by bartonjs** [comment]
Effectively, we're going to pursue providing a .NET-specific openssl.cnf, and seed it with some opinions.

**Comment by JeepNL** [comment]
**FYI:**
(_Using Server Side Blazor with Kestrel .NET Core 3.1 LTS on Linux Ubuntu 18.04 VPS with OpenSSL 1.1.1, Let's Encrypt Certificate_)

I've implemented the sample (above) from @ChrML (which is on [Kestrel web server implementation in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-3.1#listenoptionsprotocols) as well) and I am using the test suite from the Dutch Government (Websites/Email): [https://en.internet.nl/](https://en.internet.nl/)

The score is 97% and the one error I get is:

"Cipher order 
Verdict: Your web server does not enforce its own cipher preference ('I').
First found affected cipher pair:  none"

And about the sample (answer from @krwq):
> to avoid allocations might be worth storing that CipherSuitesPolicy instance somewhere - otherwise it will be constructed on every request

Is it now 'safe' the use the sample from above, as it is on [Kestrel web server implementation in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-3.1#listenoptionsprotocols)

**Comment by ChrML** [comment]
@JeepNL 

Here is an updated sample that I use in my production webserver now, that avoids unneccessary allocations:

From the Program.CreateWebHostBuilder method:
```csharp
.ConfigureKestrel((context, options) =>
{
    options.ConfigureHttpsDefaults(httpsOptions =>
    {
        if (!System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform.Windows))
        {
            httpsOptions.OnAuthenticate = (conContext, sslAuthOptions) =>
            {
                sslAuthOptions.CipherSuitesPolicy = cipherSuitesPolicy;
            };
        }
    });
});
```

A field in the Program class:
```csharp
// Ciphersuit policy for this server:
static readonly System.Net.Security.CipherSuitesPolicy cipherSuitesPolicy = new System.Net.Security.CipherSuitesPolicy
    (
        new System.Net.Security.TlsCipherSuite[]
        {
            // Cipher suits as recommended by: https://wiki.mozilla.org/Security/Server_Side_TLS
            // Listed in preferred order.

            // Highly secure TLS 1.3 cipher suits:
            System.Net.Security.TlsCipherSuite.TLS_AES_128_GCM_SHA256,
            System.Net.Security.TlsCipherSuite.TLS_AES_256_GCM_SHA384,
            System.Net.Security.TlsCipherSuite.TLS_CHACHA20_POLY1305_SHA256,

            // Medium secure compatibility TLS 1.2 cipher suits:
            System.Net.Security.TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
            System.Net.Security.TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
            System.Net.Security.TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
            System.Net.Security.TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
            System.Net.Security.TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
            System.Net.Security.TlsCipherSuite.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
            System.Net.Security.TlsCipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
            System.Net.Security.TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
        }
    );
```

**Comment by JeepNL** [comment]
@ChrML Thank you for the example. With your code I get an "A+" rating at SSL Labs. 

**FYI:** I've added some extra ciphers

```
// From: https://en.internet.nl
// High
TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
TlsCipherSuite.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
// Medium
TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,
TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384,
TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
TlsCipherSuite.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,
TlsCipherSuite.TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
TlsCipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
TlsCipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA256,
TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA,
TlsCipherSuite.TLS_DHE_RSA_WITH_AES_128_CBC_SHA256,
TlsCipherSuite.TLS_DHE_RSA_WITH_AES_128_CBC_SHA,
TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,
TlsCipherSuite.TLS_AES_128_GCM_SHA256,
TlsCipherSuite.TLS_AES_256_GCM_SHA384,
TlsCipherSuite.TLS_CHACHA20_POLY1305_SHA256
```

I've also this in **Startup.cs**

_ConfigureServices:_

```
services.AddHsts(opts =>
{
    opts.Preload = true;
    opts.IncludeSubDomains = true;
    opts.MaxAge = TimeSpan.FromDays(190);
});

```
_Configure:_
```
app.UseHsts();
app.Use(async (context, next) =>
{
    context.Response.Headers.Add("X-Frame-Options", "DENY");
    context.Response.Headers.Add("X-Content-Type-Options", "nosniff");
    context.Response.Headers.Add("X-XSS-Protection", "1; mode = block");
    context.Response.Headers.Add("Referrer-Policy", "strict-origin");
    await next();
});
```

**Comment by JeepNL** [comment]
https://en.internet.nl is a internet security test suite from the Dutch Internet Community and the Dutch Government (see: [about](https://en.internet.nl/about/)). It follows the guidelines from NCSC.nl (_National Cyber Security Centrum_) They recently updated the guidelines to comply with the newest internet standards.

With Kestrel the HTTPS test fails because Kestrel does not enforce the 'Cipher Order', it lets the browser do this (_see images_)

I hope the team can take a look at this for .NET Core 5.
![Screenshot_2019-12-13 Website test ictz net](https://user-images.githubusercontent.com/39458532/70796237-07bdad00-1da2-11ea-94aa-550686c35aa1.png)
![Screenshot_2019-12-13 Website test ictz net(1)](https://user-images.githubusercontent.com/39458532/70796245-0d1af780-1da2-11ea-8002-a03a2b54c217.png)
![Screenshot_2019-12-13 Website test ictz net(2)](https://user-images.githubusercontent.com/39458532/70796249-1015e800-1da2-11ea-885e-1507ba597458.png)
![Screenshot_2019-12-13 Website test ictz net(3)](https://user-images.githubusercontent.com/39458532/70796253-11dfab80-1da2-11ea-8253-c54d565c7595.png)

**Comment by ChrML** [comment]
@JeepNL Yes, it seems SSL Labs does not reduce the score because of lack of serverside cipher-suite preference. It does however notice and comments it, but still passes as A+.

Other stricter SSL tests are per Kestrel today impossible to pass without a reverse proxy or load balancer before, or modifying the OS OpenSSL configuration (which is often not feasible because Docker).

**Comment by JeepNL** [comment]
@ChrML I'm running it on a plain Ubuntu 18.04 VPS, no Docker. And I really want to run Kestrel stand alone, not with a reverse proxy. Do you maybe have a link to some info on how to configure OpenSSL with Kestrel so it enforces the Cipher Order on the server? I've Postfix installed, and don't get the 'Cipher Order'-error on the email test of https://en.internet.nl - It scores a 100%. It's installed on de same server.

**Comment by ChrML** [comment]
@JeepNL Sorry, never done it myself, don't know how. Maybe someone else here does. When I was googling this issue, there were other people suggesting to do that.

**Comment by JeepNL** [comment]
@chrml Found something earlier in this thread, updated my previous comment.

**Comment by JeepNL** [comment]
@bartonjs IT WORKS! 100%! I just did a copy/paste of your example and put it before the first [ header ] of "/etc/ssl/openssl.cnf". (_Server Side Blazor with Kestrel .NET Core 3.1 LTS, Ubuntu 18.04 VPS. OpenSSL 1.1.1_)

I was a bit worried first because I install the latest .NET Core via SNAP and I found a couple of more locations of openssl.conf in SNAP dirs but I just needed to edit the default (OS) one. It was really that simple. 1st try and it just worked.

Thank you all so much for your help/info.

![Screenshot_2019-12-14 Website test ictz net](https://user-images.githubusercontent.com/39458532/70847160-55e5b580-1e61-11ea-822b-bfb5e7e1bdda.png)
. Restarted my Kestrel service and tested it at https://en.internet.nl.
